static-site/src/components/ListResources/useDataFetch.ts

```diff
@@ -1,42 +1,39 @@
-import { useState, useEffect } from 'react';
-import { Group, GroupDisplayNames, PathVersions, Resource } from './types';
+import { Dispatch, SetStateAction, useState, useEffect } from 'react';
+import { Group, GroupDisplayNames, PathVersions, Resource, ResourceListingInfo } from './types';
 
 
 /**
- * Fetches the datasets for the provided `sourceId` and parses the data into an
- * array of groups, each representing a "pathogen" and detailing the available
- * resources for each, and the available versions (snapshots) for each of those
- * resources.
+ * Uses the provided callback to fetch resources and parse those
+ * resources into the `pathVersions` and `pathPrefix` values.
  *
- * The current implementation defines `versioned: boolean` for the entire API
- * response, however in the future we may shift this to the API response and it
- * may vary across the resources returned.
+ * Continues on to parse the `pathVersions`/`pathPrefix` data
+ * structures into an array of groups, each representing a "pathogen"
+ * and detailing the available resources for each, and the available
+ * versions (snapshots) for each of those resources. In the case of
+ * un-versioned resources, versions will be a zero-length array (i.e.,
+ * `[]`)
+ *
+ * The current implementation defines `versioned: boolean` for the
+ * entire API response, however in the future we may shift this to the
+ * API response and it may vary across the resources returned.
  */
-export function useDataFetch(sourceId: string, versioned: boolean, defaultGroupLinks: boolean, groupDisplayNames: GroupDisplayNames) {
+export function useDataFetch(
+  versioned: boolean,
+  defaultGroupLinks: boolean,
+  groupDisplayNames: GroupDisplayNames,
+  resourceListingCallback: (setDataFetchError:Dispatch<SetStateAction<boolean>> ) => Promise<ResourceListingInfo>,
+) : {groups: Group[] | undefined, dataFetchError: boolean} {
   const [groups, setGroups] = useState<Group[]>();
-  const [dataFetchError, setDataFetchError] = useState<boolean>();
-  useEffect(() => {
-    const url = `/list-resources/${sourceId}`;
+  const [dataFetchError, setDataFetchError] = useState<boolean>(false);
 
-    async function fetchAndParse() {
-      let pathVersions: PathVersions, pathPrefix: string;
+  useEffect((): void => {
+    async function fetchAndParse(): Promise<void> {
       try {
-        const response = await fetch(url, {headers: {accept: "application/json"}});
-        if (response.status !== 200) {
-          console.error(`ERROR: fetching data from "${url}" returned status code ${response.status}`);
-          return setDataFetchError(true);
-        }
-        ({ pathVersions, pathPrefix } = (await response.json()).dataset[sourceId]);
-      } catch (err) {
-        console.error(`Error while fetching data from "${url}"`);
-        console.error(err);
-        return setDataFetchError(true);
-      }
+        const { pathPrefix, pathVersions } = await resourceListingCallback(setDataFetchError);
```

Comment:
At the moment, if the callback _fails_ it's setting an error state and returning `undefined`, which is then causing a runtime error (because `const { pathPrefix, pathVersions } = undefined`) which gets caught and then... sets  the error state again¹ to `true`. 

Would it be simpler to just allow the callback to throw, and catch that here (which we are doing anyway)? If we want to set custom error messages then we need more sweeping changes which seem unnecessary at this stage.

¹ Conceptually speaking. In reality, I think the two `setDataFetchError` calls will be batched so you only see the last one. Maybe.

Reply:
yeah, that's fair; i've ripped out the try/catch bits in the callbacks and it's clearer. will push update soon.